### ISSUE TITLE
Create CAOM data engineering map for NIFS processed data.

### ISSUE THREAD

**Issue by ijiraq** [ijiraq/gemini_processing#6]
To distribute the outputs of the processing we will need to have the metadata stored in the CADC's caom data model.  Likely this can be largely a modification of the 'raw' caom map.

**Comment by dbohlender** [comment]
Nat, when you have a chance can you give me access to a processed NIFS science dataset so that I can inspect the headers?   Does nifty also co-add any individual calibration datasets?  If so, I'll need to see those as well.

**Comment by Nat1405** [comment]
Hi David, not a problem. [Here's a spectrum](https://www.dropbox.com/s/37s7rkx6m7ewri3/titan_final_spectrum_corrected.fits?dl=0) I happened to have lying around and I just started a reduction now to produce some cubes.

If I understand you correctly no, Nifty doesn't co-add multiple calibration datasets together. It only finds those calibrations associated with an individual program ([here's what it would see](https://archive.gemini.edu/searchform/GN-2014A-Q-85/notengineering/NotFail/not_site_monitoring/cols=CTOWEQ) for a query for program ID "GN-2014A-Q-85") and operates on those. However there is quite a bit of combining/normalizing/etc happening on those calibrations; is that what you mean?

**Comment by Nat1405** [comment]
Hi @dbohlender , I've finished a full reduction of some Titan data. [Here's a link](https://www.dropbox.com/sh/fes5m6zoxfmne0u/AADEOFvX2WWU_cBbG7VJKAjEa?dl=0). Things look pretty good except that the merging of the [final cube](https://www.dropbox.com/s/ajudzmv8pf4uvb0/TOTAL_mergedK.fits?dl=0) is a little funny (the pipeline doesn't pick up x-y offsets between observations yet). Here's an overview of the products:

![output_products_GN-2014A-Q-85](https://user-images.githubusercontent.com/17349431/81725213-5bf4b800-943a-11ea-9061-032b368bf9b5.png)

I remember you noticed it looked like some science frames were missing in the final merging? It turns out those were sky frames; the only difference between them and science frames is in their header 'POFFSET' and 'QOFFSET' values. Nifty uses those to check if the telescope is pointing more than 2 arcseconds (I think? I'd need to check) away from the target and if so flags them as a sky frame.

Long story short, we might be able to use the header information of the 'factfbrsnN.... .fits' files to figure out the temporal CAOM2 information.

**Comment by dbohlender** [comment]
Okay, good.  Offsets are much smaller for the telluric standards but that's presumably simply because they are point sources.

**Comment by Nat1405** [comment]
For sure, I would suspect most telluric standards to be observed by re-pointing the entire telescope rather than using offsets. I can look into that more though if you'd like.

**Comment by dbohlender** [comment]
No.  Obviously the telescope is repointed to the standards since they can be many degrees away from the science target.  I just meant that in the sequence of what looks like typically 5 observations of a standard, one is centered on the star and the other 4 are offset, but only an arcsecond or so vs. the >3" needed for extended targets like Titan.

**Comment by Nat1405** [comment]
Oh! Now I get it. Yes, I forget why that dithering pattern is used for tellurics but good catch. Presumably something to offset instrumental bias...

![telluric_dithering](https://user-images.githubusercontent.com/17349431/81740764-55723a80-9452-11ea-813f-89e8c06f6b8a.gif)

**Comment by Nat1405** [comment]
Just leaving this here for future reference. I was talking to Sharon and I found some containers she mentioned that might be useful for translating our .fits files to CAOM2.

https://github.com/opencadc-metadata-curation/blank2caom2

**Comment by dbohlender** [comment]
### NIFS data product provenance

Before any attempt to inject NIFS products into CAOM2 we need some decisions made about file naming and data label (observation ID) conventions and a strategy for identification of inputs and members.

Currently the file merged products in the pipeline have names TOTAL_mergedK.fits.   Presumably the K is a filter identification.   Gemini's convention in the past for co-added products (e.g. biases, flats) has been to use the root of the first file used to create the product as part of the filename, with short prefixes/suffixes added to encode the processing that has occurred.  e.g. mrgN20040513S0102_add   

So, for example, we could have a final file name like N20140505S0108total_merged_K.fits.  Perhaps the grating identifier (K in this case) can be omitted.  'total_merged_K' might instead be shortened to 'tmK' or 'tm'?

The DATALAB value in the header of the final product is not unique.  e.g. for this example it remains GN-2014A-Q-85-12-002, which I'm guessing might be the value for the first observation used to produce the product.   So this could be modified to, perhaps, GN-2014A-Q-85-12-002-TOTAL-MERGED-K or (better?) shortened to GN-2014A-Q-85-12-002-TM or -TMK.    The above GMOS-N product has a data label GN-2004A-Q-98-21-001-MRG-ADD.   This is the observation ID in CAOM2.

**Members** are not identified in a useful way in the current headers.  e.g. the above example's extension head has NCOMBINE = 20 but only three IMCMB### entries and these identify intermediate data products (e.g. merged observations from different nights) which I don't think need to be archived.  If we want to identify members in the product header it would be better to have NMEMBERS (rather than NCOMBINE) represent the number of individual unprocessed datasets used to produce the product and MEMBER## or MEMBR### contain the data labels (not file names)  of these.  (Are there ever more than 99 members?)

In a similar way, unprocessed calibration **inputs** could also be identified in the header but I'd instead suggest keywords identifying the processed flat, dark, any any other calibration files rather than individual calibration members.  These processed calibration files should then likely also be archived and have appropriate members/inputs identified in their headers in turn.

An alternative to adding this provenance info to the headers might be to modify nifty4gemini to create a parsable text file as another artifact for the data product.   We have already discussed adding the processing configuration file as an extra artifact.   However this would mean significantly more modifications to gemini2caom2 code than the FITS header route.

**Comment by ijiraq** [comment]
The MEMBR### format is a bit clunky but having the information in an external artifact makes them easy to loose.  Could there be additional extensions added to the FITS output that contained tables or similar that would provide the needed information?

**Comment by dbohlender** [comment]
Using MEMBR### would be the easiest for Sharon to code since for processed Gemini data already in the archive gemini2caom2 already uses IMCMB### entries to identify members.   (She agrees that an extra artifact is not a great alternative.)  No doubt she can do her magic with an extra extension...

**Comment by Nat1405** [comment]
Here's a draft of how a new fits extension in the products might look:

![image](https://user-images.githubusercontent.com/17349431/85057127-289dfb00-b155-11ea-9ab8-947869ce8a00.png)

![image](https://user-images.githubusercontent.com/17349431/85057157-381d4400-b155-11ea-809c-4ce8f5cf1632.png)
